iwyu.cc

```diff
@@ -2499,13 +2464,14 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
 
     const NamedDecl* decl = TypeToDeclAsWritten(type);
 
+    const char* comment = nullptr;
     // If we are forward-declarable, so are our template arguments.
     if (CanForwardDeclareType(current_ast_node()) &&
-        !IsProvidedTypeComponent(current_ast_node())) {
+        !IsProvidedTypeComponent(current_ast_node(), &comment)) {
       ReportDeclForwardDeclareUse(CurrentLoc(), decl);
       current_ast_node()->set_in_forward_declare_context(true);
     } else {
-      ReportDeclUse(CurrentLoc(), decl);
+      ReportDeclUse(CurrentLoc(), decl, comment);
```

Comment:
This new API where the comment is returned in an out-param is pretty implicit.

* IsProvidedTypeComponent is called, and if it returns false...
* ... ReportDeclUse in the else leg uses the comment it returned

I wonder if that could be simplified. I'd even prefer to do the return-type detection again to the current state.

Reply:
It could probably be slightly better if `IsProvidedTypeComponent` would return `pair<bool, const char*>` or a `IsProvidedTypeComponentReturn` structure instead of out-parameter...

Reply:
Another variant can be to return
```cpp
enum ProvidedComponentType {
  None,
  ForTypeAlias,
  ForFnReturn,
//ForAutocast
}
```
and then analyze it in `GetTypeUseComment` function (also taking `CanForwardDeclareType` result for correct handling nested types).

Reply:
I think I prefer the `pair<bool, const char*>` retval, even if it gets a little inconvenient at the call sites.

Reply:
Are you still against structured bindings?
```cpp
const auto [is_provided, comment] = IsProvidedTypeComponent();
```
looks better for me than `is_provided_comment_pair` or just `res`.

Reply:
Yeah, I agree. And it's nicer than `std::tie` because it doesn't need separate declarations.

Let's use structured bindings, I'll get used to it :)